README.rst
=======================
Python LESSCSS Compiler
=======================

The `lessc` Python package compiles content in LESS syntax into CSS by
first preprocessing it and then invoking the nodejs `lessc` program to
actually compile it. The preprocessing resolves imports which allows
the LESS content to come from anywhere, including the filesystem,
zipped EGG files, databases, etc. This solves one of the greatest
limitations of the nodejs lessc program which requires all of the LESS
dependencies to be actual files and imports to be non-absolute.

It is possible, but *quite* improbable, that this package turns into a
pure-Python implementation of a LESS compiler.

For more information on LESS, please see http://lesscss.org/.


Project
=======

* Homepage: https://github.com/canaryhealth/lessc
* Bugs: https://github.com/canaryhealth/lessc/issues


Installation
============

.. code:: bash

  $ npm install less
  $ export PATH="`pwd`/node_modules/.bin:$PATH"
  $ pip install lessc


Usage
=====

.. code:: python

  import lessc

  # compile a less asset in the "mypackage" python module:
  css = lessc.compile_asset('mypackage:path/to/style.less')

  # compile a less file on the filesystem
  css = lessc.compile_file('../path/to/style.less')

  # compile a less file-like object
  # (if there are any relative imports, specifying `base` is necessary)
  css = lessc.compile_file(open('../path/to/style.less'))

  # compile a less string
  # (if there are any relative imports, specifying `base` is necessary)

  less = '''
    @import "mypackage:path/to/style.less";
    // ...other less statements...
  '''

  css = lessc.compile(less)


For more import resolution control, you can specify a custom URI
resolver (which resolves a relative URI to an absolute URI) and a
custom URI loader (which loads an absolute URI), using the `resolver`
and `loader` keywords to any compile* function:

.. code:: python

  import lessc

  def resolver(uri, base=None):
    '''
    Returns the absolute URI given a base URI and a potentially
    absolute or relative URI. Note that `base` may be ``None`` when
    the initial URI is being resolved (depending on the Compiler
    configuration).
    '''
    # calculate the absolute URI...
    return absolute_uri

  def loader(uri):
    '''
    Returns a file-like object that will return the content of
    the specified URI.
    '''
    # fetch the object defined by `uri`
    return file_like_object

  css = lessc.compile_asset(uri, resolver=resolver, loader=loader)


Or you can create a custom compiler and set it as the default compiler:

.. code:: python

  import lessc

  compiler = lessc.Compiler(resolver=my_custom_resolver, loader=my_custom_loader)
  lessc.set_default_compiler(compiler)

  # this will now use `my_custom_resolver` and `my_custom_loader`
  # to compile asset `uri`
  css = lessc.compile_asset(uri)


Limitations
===========

Currently, there exist the following restrictions in what kind of
LESS syntax is used:

* Recursive imports are not supported and will be silently ignored.

* Only the following import keywords are explicitly supported:

  * ``less``
  * ``css``
  * ``once``
  * ``optional``

  The keywords ``reference``, ``inline``, and ``multiple`` are NOT
  supported. Any other keywords may or may not be supported (since
  only the above keywords were documented as of this writing,
  2015/06/26).


CHANGELOG.rst
=========
ChangeLog
=========


v0.1.3
======

* Added "simplistic" support for commented-out ``@import`` statements
* Initial move to use `fso` for testing
* Added ignoring of node's `util.print` DeprecationWarning


v0.1.2
======

* Removed `distribute` dependency
* Updated documentation


v0.1.0
======

* First release


lessc/test.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: Philip J Grabner <phil@canary.md>
# date: 2015/06/26
# copy: (C) Copyright 2015-EOT Canary Health, Inc., All Rights Reserved.
#------------------------------------------------------------------------------

import os.path
import unittest

import six
import asset
import fso

from .compiler import Compiler, urijoin, CompileError, AssetError

#------------------------------------------------------------------------------
class TestCompiler(unittest.TestCase):

  #----------------------------------------------------------------------------
  def test_urijoin(self):
    self.assertEqual(urijoin(None, '/path/to/foo'), 'file:///path/to/foo')
    self.assertEqual(urijoin(None, 'path/to/foo'), 'file://' + os.path.abspath('path/to/foo'))
    self.assertEqual(urijoin(None, 'http://foo.com/a/b'), 'http://foo.com/a/b')
    self.assertEqual(urijoin(None, 'pkg:a/b'), 'pkg:a/b')
    self.assertEqual(urijoin('file:///a/b', '/c/d'), 'file:///c/d')
    self.assertEqual(urijoin('file:///a/b', 'c/d'), 'file:///a/c/d')
    self.assertEqual(urijoin('http://foo.com/a/b', '/c/d'), 'http://foo.com/c/d')
    self.assertEqual(urijoin('http://foo.com/a/b', 'c/d'), 'http://foo.com/a/c/d')
    self.assertEqual(urijoin('pkg:a/b', '/c/d'), 'pkg:c/d')
    self.assertEqual(urijoin('pkg:a/b', 'c/d'), 'pkg:a/c/d')
    self.assertEqual(urijoin('http://foo.com/a/b', 'file:///c/d'), 'file:///c/d')
    self.assertEqual(urijoin('file:///c/d', 'http://foo.com/a/b'), 'http://foo.com/a/b')
    self.assertEqual(urijoin('file:///c/d', 'pkg:a/b'), 'pkg:a/b')

  #----------------------------------------------------------------------------
  def test_invalid_syntax(self):
    with self.assertRaises(CompileError):
      Compiler().compile('this is not valid less.\n')

  #----------------------------------------------------------------------------
  def test_invalid_import(self):
    with self.assertRaises(AssetError):
      Compiler().compile('@import "no-such-asset";\n')

  #----------------------------------------------------------------------------
  def test_commentedOutImport_simple(self):
    with fso.push() as overlay:
      with open('black.less', 'wb') as fp:
        fp.write('.b { color: #000; }')
      with open('white.less', 'wb') as fp:
        fp.write('.w { color: #fff; }')
      self.assertEqual(
        Compiler().compile('''
          @import "black.less";
          // @import "no-such-asset";
          @import "white.less";
        '''),
        '.b{color:#000;}\n.w{color:#fff;}\n')

  #----------------------------------------------------------------------------
  def test_commentedOutImport_interleaved(self):
    with fso.push() as overlay:
      with open('black.less', 'wb') as fp:
        fp.write('.b { color: #000; }')
      with open('white.less', 'wb') as fp:
        fp.write('.w { color: #fff; }')
      self.assertEqual(
        Compiler().compile('''
          // @import "no-such-asset";
          @import "black.less";
          .r { color: #f00; } // @import "another-bad-asset"; .g { color: #0f0; }
          @import "white.less";
        '''),
        '.b{color:#000;}\n.r{color:#f00;}\n.w{color:#fff;}\n')

  #----------------------------------------------------------------------------
  def test_string(self):
    self.assertEqual(
      Compiler().compile('@color: #000; .black { color: @color; }'),
      '.black{color:#000000;}\n')

  #----------------------------------------------------------------------------
  def test_mixed_import(self):
    self.assertEqual(
      Compiler().compile('@import "lessc:res/test-03.less";'),
      asset.load('lessc:res/test-03.css').read())

  #----------------------------------------------------------------------------
  def test_asset_simple(self):
    self.assertEqual(
      Compiler().compile_asset('lessc:res/test-01.less'),
      asset.load('lessc:res/test-01.css').read())

  #----------------------------------------------------------------------------
  def test_file_relative_import(self):
    self.assertEqual(
      Compiler().compile_asset('lessc:res/test-02.less'),
      asset.load('lessc:res/test-02.css').read())

  #----------------------------------------------------------------------------
  def test_extension_defaulting(self):
    self.assertEqual(
      Compiler().compile_asset('lessc:res/test-03.less'),
      asset.load('lessc:res/test-03.css').read())

  #----------------------------------------------------------------------------
  def test_importOption_once_explicit(self):
    def resolver(uri, base=None):
      return uri
    def loader(uri):
      if uri == 'red.less':
        return six.StringIO('.red { color: red; }')
      raise NotImplementedError(uri)
    self.assertEqual(
      Compiler(resolver=resolver, loader=loader).compile('''
        @import (once) "red";
        @import (once) 'red';
      '''),
      '.red{color:red;}\n')

  #----------------------------------------------------------------------------
  def test_importOption_once_implicit(self):
    def resolver(uri, base=None):
      return uri
    def loader(uri):
      if uri == 'red.less':
        return six.StringIO('.red { color: red; }')
      raise NotImplementedError(uri)
    self.assertEqual(
      Compiler(resolver=resolver, loader=loader).compile('''
        @import "red";
        @import 'red';
      '''),
      '.red{color:red;}\n')

  #----------------------------------------------------------------------------
  def test_importOption_css_explicit(self):
    # NOTE: this is only supported by lessc >= 1.4.0
    try:
      self.assertEqual(
        Compiler().compile('@import (css) "path.ext";'),
        '@import "path.ext";\n')
    except CompileError:
      with self.assertRaises(CompileError) as cm:
        Compiler().compile('@import (css) "path.ext";')
      self.assertIn('ParseError: Syntax Error on line 1', str(cm.exception))
      raise unittest.SkipTest('lessc >= 1.4.0 required for "import (css) ..."')

  #----------------------------------------------------------------------------
  def test_importOption_css_implicit(self):
    self.assertEqual(
      Compiler().compile('@import "path.css";'),
      '@import "path.css";\n')

  #----------------------------------------------------------------------------
  def test_importOption_less(self):
    def resolver(uri, base=None):
      return uri
    def loader(uri):
      if uri == 'red.css':
        return six.StringIO('.red { color: red; }')
      raise NotImplementedError(uri)
    self.assertEqual(
      Compiler(resolver=resolver, loader=loader).compile('@import (less) "red.css";'),
      '.red{color:red;}\n')


#------------------------------------------------------------------------------
# end of $Id$
# $ChangeLog$
#------------------------------------------------------------------------------


lessc/__init__.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: Philip J Grabner <phil@canary.md>
# date: 2015/06/26
# copy: (C) Copyright 2015-EOT Canary Health, Inc., All Rights Reserved.
#------------------------------------------------------------------------------

from .compiler import Compiler

#------------------------------------------------------------------------------

_default_compiler = None

#------------------------------------------------------------------------------
def get_default_compiler():
  global _default_compiler
  if _default_compiler is None:
    _default_compiler = Compiler()
  return _default_compiler

#------------------------------------------------------------------------------
def set_default_compiler(compiler):
  global _default_compiler
  _default_compiler = compiler

#------------------------------------------------------------------------------
def get_compiler(resolver=None, loader=None):
  if resolver is None and loader is None:
    return get_default_compiler()
  return Compiler(resolver=resolver, loader=loader)

#------------------------------------------------------------------------------
def compile(target, **kw):
  return get_compiler(**kw).compile(target)

#------------------------------------------------------------------------------
def compile_file(target, **kw):
  return get_compiler(**kw).compile_file(target)

#------------------------------------------------------------------------------
def compile_asset(target, **kw):
  return get_compiler(**kw).compile_asset(target)

#------------------------------------------------------------------------------
# end of $Id$
# $ChangeLog$
#------------------------------------------------------------------------------


lessc/compiler.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: Philip J Grabner <phil@canary.md>
# date: 2015/06/26
# copy: (C) Copyright 2015-EOT Canary Health, Inc., All Rights Reserved.
#------------------------------------------------------------------------------

import os
import os.path
import re
import subprocess

import six
from six.moves.urllib import parse as urlparse
import asset

#------------------------------------------------------------------------------

assetspec_cre = re.compile(r'^[a-z_][a-z0-9_]*:[^/]', re.IGNORECASE)
import_cre    = re.compile(
  r'@import\s*(?:\((?P<options>[^)]*)\))?\s*(?P<qchar>["\'])(?P<uri>.*?)(?P=qchar)\s*;')

utilprintwarning_cre = re.compile(
  r'\(node:\d+\) DeprecationWarning: util.print is deprecated. Use console.log instead.')

class LesscError(Exception): pass
class AssetError(LesscError):
  def __init__(self, cause=None, *args, **kw):
    super(AssetError, self).__init__(*args, **kw)
    self.cause = cause
class CompileError(LesscError): pass
class UnsupportedError(LesscError): pass

#------------------------------------------------------------------------------
def urijoin(base, uri):
  # todo: ugh. this is trying way too hard... perhaps abandon support
  #       for asset-specs?...
  if uri.startswith('/'):
    if base is None or base.startswith('file:///'):
      return 'file://' + uri
    if not uri.startswith('//') and assetspec_cre.match(base):
      return base.split(':', 1)[0] + ':' + uri[1:]
    return urlparse.urljoin(base, uri)
  if uri.startswith('file:///'):
    return uri
  if assetspec_cre.match(uri):
    return uri
  if base is None:
    if urlparse.urlparse(uri).scheme:
      return uri
    return 'file://' + os.path.abspath(uri)
  if not assetspec_cre.match(base):
    return urlparse.urljoin(base, uri)
  parts = base.split(':', 1)
  return parts[0] + ':' + urlparse.urljoin(parts[1], uri)

#------------------------------------------------------------------------------
def resolver(uri, base=None):
  return urijoin(base, uri)

#------------------------------------------------------------------------------
def loader(uri):
  if uri.startswith('file://'):
    # todo: what about closing this file descriptor?... let the gc do it?
    return open(uri[len('file://') : ], 'rb')
  if assetspec_cre.match(uri):
    return asset.load(uri).peek()
  raise UnsupportedError('no asset loader for URI %r' % (uri,))

#------------------------------------------------------------------------------
class Compiler(object):

  lessc_exec = 'lessc'

  #----------------------------------------------------------------------------
  def __init__(self, base_uri=None, resolver=resolver, loader=loader, *args, **kw):
    super(Compiler, self).__init__(*args, **kw)
    self.base_uri = base_uri
    self.resolver = resolver
    self.loader   = loader

  #----------------------------------------------------------------------------
  def compile_asset(self, spec):
    return self._compile(spec, asset.load(spec))

  #----------------------------------------------------------------------------
  def compile_file(self, filename):
    filename = os.path.abspath(filename)
    return self._compile('file://' + filename, open(filename, 'rb'))

  #----------------------------------------------------------------------------
  def compile_uri(self, uri):
    uri = self.resolver(uri, base=self.base_uri)
    return self._compile(uri, self.loader(uri))

  #----------------------------------------------------------------------------
  def compile(self, target):
    '''
    Compiles `target` from LESS to CSS. `target` can be one of the following:

    * A unicode string object containing the less content.

    * A binary data object containing the encoded less content (an
      attempt will be made to decode this into a unicode string).

    * A file-like object that will return the less content when read.
    '''
    if isinstance(target, six.binary_type):
      try:
        target = target.decode()
      except UnicodeDecodeError:
        try:
          target = target.decode('utf-8')
        except UnicodeDecodeError:
          raise ValueError('binary input could not be converted to unicode')
    if isinstance(target, six.string_types):
      return self._compile(None, six.StringIO(target))
    return self._compile(None, target)

  #----------------------------------------------------------------------------
  def _compile(self, uri, source):
    output = six.StringIO()
    self._preprocess(uri, source, output, dict())
    output = output.getvalue().encode('utf-8')
    output = self._less2css(output)
    return output

  #----------------------------------------------------------------------------
  def _less2css(self, data):
    proc = subprocess.Popen(
      [os.environ.get('LESSC', self.lessc_exec), '-x', '-'],
      stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    output, err = proc.communicate(data)
    if err and utilprintwarning_cre.match(err.strip()):
      err = None
    if err:
      raise CompileError(err)
    return output

  #----------------------------------------------------------------------------
  def _preprocess(self, uri, source, output, state):
    if 'imported' not in state:
      state['imported'] = []
    data = source.read()
    def replacer(match):
      # TODO: *** hack alert *** this is embarrassing... this should
      #       not be here. it *should* be done wherever the data
      #       stream is actually being parsed. but currently, that is
      #       in the external `lessc` node program. ugh. and this is
      #       such a simplistic, error-prone implementation. i'm so
      #       sorry. for example, it would incorrectly handle::
      #          .f { background: url("https://foo"); } @import "bar";
      #       fortunately, i expect that to be rare.
      #       unless the source is already css-minified.
      #       don't do that.
      befline = match.string[ : match.start() ].rsplit('\n', 1)[-1]
      if '//' in befline:
        return ''
      options = match.group('options') or 'once'
      options = [i.strip().lower() for i in options.split(',')]
      for opt in options:
        if opt in ('reference', 'inline', 'multiple'):
          raise UnsupportedError(
            'import option %r is not supported in python lessc' % (opt,))
      suburi = match.group('uri')
      if not os.path.splitext(suburi)[1]:
        suburi += '.less'
      if 'css' in options or ( suburi.endswith('.css') and 'less' not in options ):
        return match.string[match.start() : match.end()]
      suburi = self.resolver(suburi, base=uri)
      if suburi in state['imported']:
        return ''
      state['imported'].append(suburi)
      try:
        subsrc = self.loader(suburi)
      except (IOError, asset.NoSuchAsset) as exc:
        if 'optional' in options:
          return ''
        raise AssetError(cause=exc)
      subout = six.StringIO()
      self._preprocess(suburi, subsrc, subout, state)
      return subout.getvalue()
    data = import_cre.sub(replacer, data)
    output.write(data)


#------------------------------------------------------------------------------
# end of $Id$
# $ChangeLog$
#------------------------------------------------------------------------------
